apps/glusterfs/health.go

```diff
@@ -0,0 +1,123 @@
+package glusterfs
+
+import (
+	"bytes"
+	"encoding/gob"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/heketi/heketi/pkg/glusterfs/api"
+
+	"github.com/boltdb/bolt"
+	"github.com/lpabon/godbc"
+)
+
+var clusterHealth = "ClusterHealth"
+
+type dbchecker struct {
+	CurrentTime time.Time
+}
+
+func NewHealthz() *api.Healthz {
+	health := &api.Healthz{}
+	health.NodesStatus = make([]api.NodeHealthStatus, 0)
+	return health
+}
+
+func (d *dbchecker) Marshal() ([]byte, error) {
+	var buffer bytes.Buffer
+	enc := gob.NewEncoder(&buffer)
+	err := enc.Encode(*d)
+
+	return buffer.Bytes(), err
+}
+
+func (d *dbchecker) Unmarshal(buffer []byte) error {
+	dec := gob.NewDecoder(bytes.NewReader(buffer))
+	return dec.Decode(d)
+}
+
+func (*dbchecker) BucketName() string {
+	return BOLTDB_BUCKET_DBATTRIBUTE
+}
+
+func (d *dbchecker) Write(tx *bolt.Tx) error {
+	godbc.Require(tx != nil)
+	return EntrySave(tx, d, clusterHealth)
+}
+
+func (d *dbchecker) Read(tx *bolt.Tx) error {
+	godbc.Require(tx != nil)
+	return EntryLoad(tx, d, clusterHealth)
+}
+
+func (a *App) Healthz(w http.ResponseWriter, r *http.Request) {
+	info := NewHealthz()
+
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+
+	//db Read check
+	err := a.db.View(func(tx *bolt.Tx) error {
+		h := dbchecker{}
+		return h.Read(tx)
+	})
+
+	if err != nil && err != ErrNotFound {
+		info.ClusterHeath = api.ClusterFaulty.String()
+		info.DbState = api.DbFaulty.String()
+		w.WriteHeader(http.StatusInternalServerError)
+		if err := json.NewEncoder(w).Encode(info); err != nil {
+			panic(err)
+		}
+		return
+	}
+	info.DbState = api.DbRead.String()
+
+	//db write check
+	a.db.Update(func(tx *bolt.Tx) error {
+		h := dbchecker{CurrentTime: time.Now()}
+		return h.Write(tx)
+	})
+	if err == nil {
+		info.DbState = api.DbWrite.String()
+	}
+	//cluster status check
+	err = a.db.View(func(tx *bolt.Tx) error {
+		health := NewNodeHealthCache(0, 0, a.db, a.executor)
```

Comment:
Don't create a new health status cache. We should re-use the one previously set up when the app is set up. If the cache is disabled, simply report an empty but non-failure result.

Reply:
sorry I missed that we  are storing  the heathcache in app, updated code